### ISSUE TITLE
Migration for OperatingSystemManager - Test Release 2.22

### ISSUE THREAD

**Issue by ahmedwaleedmalik** [kubermatic/kubermatic#11748]
## Instructions
On KKP 2.21 installation, do the following:
1. Create 2-3 user clusters
2. Create an OSP in one of those cluster's user-cluster namespace. You can copy https://github.com/kubermatic/operating-system-manager/blob/main/deploy/osps/default/osp-ubuntu.yaml and rename it since `osp-usercluster` is a default OSP which OSM will create by itself.
3. Create an OSP in the seed namespace i.e. kubermatic.  You can copy https://github.com/kubermatic/operating-system-manager/blob/main/deploy/osps/default/osp-ubuntu.yaml and rename it to `osp-seed`.
4. Ensure that `osp-seed` is copied to all the user cluster namespaces.
5. Upgrade to KKP 2.22
6. Make sure there are no errors in the logs regarding these migrations.
7. Ensure that all OSPs and OSCs have been removed from the seed cluster.
8. `osp-seed` in seed namespace should have been converted to `CustomOperatingSystemProfile` from `OperatingSystemProfile`.
9. Check user cluster to make sure that `osp-usercluster` is there in `kube-system` namespace.
10. Ensure that OSP and OSC CRDs have been removed from the seed cluster.

#### Special case for KubeOne

For this case we can follow everything from the previous mentioned instructions with the following modifications:
1. Before upgrading to KKP 2.22, create a deployment named `operating-system-manager` in the `kube-system` namespace in seed. You can use https://raw.githubusercontent.com/kubernetes/website/main/content/en/examples/controllers/nginx-deployment.yaml
2. At step 10, ensure that OSP and OSC CRDs are **NOT** removed.

## References
Ticket: https://github.com/kubermatic/kubermatic/issues/11679
PR: https://github.com/kubermatic/kubermatic/pull/11720

**Comment by xmudrii** [comment]
> Ensure that all OSPs and OSCs have been removed from the seed cluster.

```
Error from server (NotFound): Unable to list "operatingsystemmanager.k8c.io/v1alpha1, Resource=operatingsystemprofiles": the server could not find the requested resource (get operatingsystemprofiles.operatingsystemmanager.k8c.io)

Error from server (NotFound): Unable to list "operatingsystemmanager.k8c.io/v1alpha1, Resource=operatingsystemconfigs": the server could not find the requested resource (get operatingsystemconfigs.operatingsystemmanager.k8c.io)
```

> osp-seed in seed namespace should have been converted to CustomOperatingSystemProfile from OperatingSystemProfile.

```
$ k get customoperatingsystemprofiles.operatingsystemmanager.k8c.io -n kubermatic
NAME       AGE
osp-seed   43m
```

> Check user cluster to make sure that osp-usercluster is there in kube-system namespace.

```
NAME       AGE
osp-seed   30m
osp-uc     11m
```

**Comment by embik** [comment]
Note on the KubeOne case:

> Before upgrading to KKP 2.22, create a deployment named operating-system-manager in the kube-system namespace in seed

This deployment already exists in the seed when installed with KubeOne. I assume the step was meant to simulate a KubeOne setup, which isn't necessary when actually installed on top of KubeOne. Skipping it.

**Comment by embik** [comment]
I've verified the migration and while the seed-controller-manager logs that it cannot find OSPs for quite some time, that seems to be a temporary issue until CRDs are applied to user clusters. No other issues were observed, I was able to verify that OSPs were migrated into `CustomOperatingSystemProfiles` and into OSPs on user clusters, that custom OSPs on the Seed were cleaned up, etc. Everything worked according to the given instructions.

Closing.
